### PR TITLE
Disassembler arch for RX microcontroller (MCU) family 

### LINKED ISSUE #24: Renesas RX architecture support
The RX microcontroller (MCU) family uses a 32-bit enhanced Harvard CISC architecture

- https://www.renesas.com/us/en/products/microcontrollers-microprocessors/rx-32-bit-performance-efficiency-mcus/rx-features
- [Renesas_RX63N_Chapter03_Slides_Ver0.1.pdf](https://github.com/rizinorg/ideas/files/7755849/Renesas_RX63N_Chapter03_Slides_Ver0.1.pdf)
- [RX Family RXv1 Instruction Set Architecture.pdf](https://github.com/rizinorg/ideas/files/7755853/REN_r01us0032ej0130-rxv1sm_MAS_20191226.pdf)
- [RX Family RXv2 Instruction Set Architecture.pdf](https://github.com/rizinorg/ideas/files/7755895/REN_r01us0071ej0100_rxv2sm_MAS_20131112.pdf)
- RX Family RXv3 Instruction Set Architecture
- [Ghidra plugin](https://github.com/redballoonsecurity/rx-proc-ghidra)

There is LLVM port for RX https://lists.llvm.org/pipermail/llvm-dev/2020-April/140546.html
Thus, it's theoretically possible to use the [auto-sync procedure](https://github.com/capstone-engine/capstone/issues/2015) to add its support to capstone (see Tricore support in capstone as a good example).

It is also supported by QEMU: https://gitlab.com/qemu-project/qemu/-/tree/master/target/rx

### PR DESCRIPTION
**Pull request by Heersin**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Introduction
- https://www.renesas.com/us/en/products/microcontrollers-microprocessors/rx-32-bit-performance-efficiency-mcus/rx-features
- [Ghidra plugin](https://github.com/redballoonsecurity/rx-proc-ghidra)

Datasheet sources:
- [RX Family RXv1 Instruction Set Architecture.pdf](https://github.com/rizinorg/ideas/files/7755853/REN_r01us0032ej0130-rxv1sm_MAS_20191226.pdf)

Done:
- [x] Rxv1 disasm
- [x] Rxv1 analysis
- [x] Rxv1 test
- [ ] ...

**Test plan**

- [x] test binary 
  - db/asm/rx
  - db/analysis/rx

**known issues**
1. disassembler prefetch some bytes from buffer, when using `pd` to disassembly large amount of bytes,  rizin default buffer may filled with zeroes while binary has unread data.
2. invalid segment warning, maybe rx-gcc warning?

**Closing issues**

Closes https://github.com/rizinorg/ideas/issues/24

### CODE REVIEW

File: librz/asm/arch/rx/rx.c [hunk 1]
```diff
@@ -0,0 +1,149 @@
+#include "rx.h"
```
Comment (XVilka) [inline librz/asm/arch/rx/rx.c:4]: Missing SPDX

File: librz/asm/arch/rx/rx_datasheet.c [hunk 2]
```diff
@@ -0,0 +1,300 @@
+#include "rx_datasheet.h"
```
Comment (XVilka) [inline librz/asm/arch/rx/rx_datasheet.c]: SPDX, and other files too

**Review by XVilka — CHANGES_REQUESTED** [review]
Please add disassembly instruction-level tests before starting working on the analysis plugin and tests.

**Comment by Heersin** [comment]
> Please add disassembly instruction-level tests before starting working on the analysis plugin and tests.

done

File: librz/analysis/p/analysis_rx.c [hunk 3]
```diff
@@ -0,0 +1,255 @@
+// SPDX-FileCopyrightText: 2024 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+#include <asm/arch/rx/rx.h>
```
Comment (XVilka) [inline librz/analysis/p/analysis_rx.c:4]: One empty line between SDPX and includes please

File: librz/asm/arch/rx/rx.c [hunk 4]
```diff
@@ -0,0 +1,163 @@
+// SPDX-FileCopyrightText: 2024 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+#include "rx.h"
```
Comment (XVilka) [inline librz/asm/arch/rx/rx.c:4]: Same, empty line please

File: librz/asm/arch/rx/rx.c [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+		if (opr->v.reg.as_base) {
+			rz_strbuf_appendf(buf, "%s,%s",
+				RxNameReg(opr->v.reg.ri),
+				RxNameReg(opr->v.reg.reg));
+		} else {
+			rz_strbuf_appendf(buf, "%s", RxNameReg(opr->v.reg.reg));
+		}
+
+		if (opr->v.reg.fix_mode == RX_FIXOP_POST_INC) {
+			rz_strbuf_appendf(buf, "+");
```
Comment (XVilka) [inline librz/asm/arch/rx/rx.c]: If you append static string, use `rz_strbuf_append()`

File: librz/asm/arch/rx/rx.c [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+				RxNameOp(inst->op));
+		}
+	} else {
+		rz_strbuf_appendf(buf, "%s", RxNameOp(inst->op));
+	}
+
+	if (inst->sz_mark != RX_EXT_NON) {
+		rz_strbuf_appendf(buf, ".%s ", RxNameExt(inst->sz_mark));
+	} else {
+		rz_strbuf_appendf(buf, " ");
```
Comment (XVilka) [inline librz/asm/arch/rx/rx.c]: Same

File: librz/asm/arch/rx/rx.c [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+/**
+ * Parse binary data to RxInst according to RxDesc
+ * \param inst
+ * \param bytes_read
+ * \param buf
+ * \param buf_len
+ * \return
+ */
+RZ_API bool rx_dis(RxInst RZ_OUT *inst, st32 RZ_OUT *bytes_read, const ut8 *buf, size_t buf_len) {
```
Comment (XVilka) [inline librz/asm/arch/rx/rx.c]: Add `rz_return..` assert and `RZ_NONNULL` for `inst` and `bytes_read`, maybe `buf`?

File: librz/asm/arch/rx/rx.c [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	if (has_opr2) {
+		rz_strbuf_appendf(buf, "%s", opr2_buf.buf);
+	}
+
+	return true;
+}
+
+/**
+ * Parse binary data to RxInst according to RxDesc
+ * \param inst
```
Comment (XVilka) [inline librz/asm/arch/rx/rx.c]: Missing description of the parameters

File: librz/asm/arch/rx/rx_datasheet.c [hunk 9]
```diff
@@ -0,0 +1,306 @@
+// SPDX-FileCopyrightText: 2024 heersin <teablearcher@gmail.com>
```
Comment (XVilka) [inline librz/asm/arch/rx/rx_datasheet.c]: I am not sure `datasheet` is a good file name. Maybe `opcodes` is better?

File: librz/asm/arch/rx/rx_datasheet.c [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+// SPDX-License-Identifier: LGPL-3.0-only
+#include "rx_datasheet.h"
```
Comment (XVilka) [inline librz/asm/arch/rx/rx_datasheet.c]: Same

File: librz/asm/arch/rx/rx_datasheet.h [hunk 11]
```diff
@@ -0,0 +1,280 @@
+// SPDX-FileCopyrightText: 2024 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+#ifndef RX_DATASHEET_H
```
Comment (XVilka) [inline librz/asm/arch/rx/rx_datasheet.h]: Same

File: librz/asm/arch/rx/rx_inst.c [hunk 12]
```diff
@@ -0,0 +1,596 @@
+// SPDX-FileCopyrightText: 2024 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+#include "rx_inst.h"
```
Comment (XVilka) [inline librz/asm/arch/rx/rx_inst.c:4]: Same

File: librz/asm/arch/rx/rx_str.inc [hunk 13]
```diff
@@ -0,0 +1,170 @@
+// SPDX-FileCopyrightText: 2024 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+const char *opnames[_RX_OP_COUNT] = {
```
Comment (XVilka) [inline librz/asm/arch/rx/rx_str.inc:4]: Same

File: librz/asm/p/asm_rx.c [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+#include <rz_lib.h>
+#include <rz_asm.h>
+#include <asm/arch/rx/rx.h>
+
+static int disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
+	RxInst inst = { 0 };
+	st32 bytes_read;
+
+	if (!rx_dis(&inst, &bytes_read, buf, len)) {
+		rz_strbuf_set(&op->buf_asm, "(invalid)");
```
Comment (XVilka) [inline librz/asm/p/asm_rx.c]: You can use the new function `rz_asm_op_set_asm()`

File: test/db/analysis/rx [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+|   | ,===< 0xfff40193      bleu.B  #0x2c, 
+|   | |:|   0xfff40195      mov.L [r10], r5, 
+|   | |:|   0xfff40197      mov.B [r5], r5, 
+|   | |:|   0xfff40199      movu.B r5, r5, 
+|   | |:|   0xfff4019b      cmp   #IMM:8(#0x5a), r5, 
+|   |,====< 0xfff4019e      bgtu.B  #0x21, 
+|   |||:|   0xfff401a0      mov.L [r10], r5, 
+|   |||:|   0xfff401a2      mov.B [r5], r5, 
+|   |||:|   0xfff401a4      movu.B r5, r5, 
+|   |||:|   0xfff401a6      add   #IMM:8(#0xbf), r5, r5
```
Comment (XVilka) [inline test/db/analysis/rx]: I am not sure about this syntax, though; we tend to simplify this stuff. @wargio @Rot127  @ret2libc @thestr4ng3r @imbillow, do you have any opinion on that?

**Review by XVilka — CHANGES_REQUESTED** [review]
Good job!
Please update also tests where plugins are listed:
```
[XX] db/cmd/cmd_list Print the asm/analysis plugins in quiet mode
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc Laq =
-- stdout
--- expected
+++ actual
@@ -36,6 +36,7 @@
 pyc
 rl78
 rsp
+rx
 sh
 snes
 sparc



[XX] db/cmd/cmd_list Print the asm/analysis plugins
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc La =
-- stdout
--- expected
+++ actual
@@ -36,6 +36,7 @@
 _dA_  8 16       pyc         LGPL3   PYC disassemble plugin
 adA_  32         rl78        LGPL3   Renesas RL78 disassembler (by Bastian Engel)
 _dA_  32         rsp         LGPL3   Reality Signal Processor
+_dA_  32         rx          LGPL3   Renesas RX Family disassembler (by Heersin)
 adAe  32         sh          LGPL3   SuperH-4 CPU (by DMaroo)
 _dA_  8 16       snes        LGPL3   SuperNES CPU
 _dA_  32 64      sparc       BSD     Capstone SPARC disassembler
 ```

File: test/db/analysis/rx (hunk 15 above)
Comment (Heersin) [reply test/db/analysis/rx]: or keep the same syntax with rx-objdump ?

File: librz/asm/arch/rx/rx.c (hunk 7 above)
Comment (Heersin) [reply librz/asm/arch/rx/rx.c]: marked, will fix it

File: librz/asm/arch/rx/rx_datasheet.c (hunk 9 above)
Comment (Heersin) [reply librz/asm/arch/rx/rx_datasheet.c]: what about opcode_detail or opcode_desc ?

Comment (XVilka) [reply librz/asm/arch/rx/rx_datasheet.c]: Yes, that would be ok

File: test/db/analysis/rx (hunk 15 above)
Comment (wargio) [reply test/db/analysis/rx]: let's use something that makes sense like `add 0xbf, r5, r5`

Comment (Rot127) [reply test/db/analysis/rx]: It this the syntax of rx-objdump/ISA? Generally I vote for more similarity with the standard tooling/being as close to the ISA as possible. Makes it easier for people to transfer their knowledge and habits when using Rizin.
And if we do simplifications, they should be optional.
Deviating too much from the ISA syntax might makes more trouble in the future.

Comment (XVilka) [reply test/db/analysis/rx]: We already diverge from objdump in many other architectures in favor of readability. Thus I am for easier representation.

Comment (Rot127) [reply test/db/analysis/rx]: Simplifications are definitely fine, though would prefer to have them in a way, that we can make them optional easily.
Meaning, it should be only a matter of one or two additional if-clauses to go back to standard ISA syntax.

Comment (Heersin) [reply test/db/analysis/rx]: well, I will change it to simple syntax likes `add num, r5, r5`

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
@Heersin I merged it as is, to include the basic support of RX in the release, but the analysis needs to be improved a lot. That can be done in a separate PR.

**Pull request merged**